<H1><B>PANDAS</B></H1>
Pandas is a package for data manipulation and analysis in Python. The name Pandas is derived from the econometrics term Panel Data. Pandas incorporates two additional data structures into Python, namely <b>Pandas Series</b> and <b>Pandas DataFrame</b>. These data structures allow us to work with labeled and relational data in an easy and intuitive manner. 

## **Why Use Pandas?**
The recent success of machine learning algorithms is partly due to the huge amounts of data that we have available to train our algorithms on. However, when it comes to data, quantity is not the only thing that matters, the quality of your data is just as important.

More often than not, large datasets will often have missing values, outliers, incorrect values, etc… Having data with a lot of missing or bad values, for example, is not going to allow your machine learning algorithms to perform well.

This is where Pandas come in. Pandas Series and DataFrames are designed for fast data analysis and manipulation, as well as being flexible and easy to use. Below are just a few features that makes Pandas an excellent package for data analysis:

- Allows the use of labels for rows and columns
- Easy handling of NaN values
- Is able to load data of different formats into DataFrames
- Can join and merge different datasets together
- It integrates with NumPy and Matplotlib

For these and other reasons, Pandas DataFrames have become one of the most commonly used Pandas object for data analysis in Python.

## **PANDAS SERIES:**
A Pandas series is a one-dimensional array-like object that can hold many data types, such as numbers or strings. One of the main differences between Pandas Series and NumPy ndarrays is that you can assign an index label to each element in the Pandas Series. In other words, you can name the indices of your Pandas Series anything you want. Another big difference between Pandas Series and NumPy ndarrays is that Pandas Series can hold data of different data types.

In [ ]:
# We first have to import pandas using import statement in python

import pandas as pd  

Let's create a pandas series.  For that we just write <B>pd.Series()</b>

```
items = pd.Series(data, index)    # For now we just names the series object in a variable named item
```



So while creating series we can pass arguments for <u>data</u> and <u>index</u>

In [ ]:
item = pd.Series(data=[15, 5, 'No'], index=['chocolates', 'chips', 'milk'])
item

chocolates    15
chips          5
milk          No
dtype: object

As we see the series is displayed as indicies in the first column and data in the second column.

<br><br>
## **Attributes of Panda Series:**

Let's see some of the attributes of the pandas series that helps us to understand our series

In [ ]:
item.shape   # gives us the  sizes of each dimension of the data

(3,)

In [ ]:
item.ndim   # gives us the number of dimensions of the data

1

In [ ]:
item.size   # gives us the total number of items in the array 

3

In [ ]:
item.index   # gives us the list indeices of the series 

Index(['chocolates', 'chips', 'milk'], dtype='object')

In [ ]:
item.values    #gives us the data of the series

array([15, 5, 'No'], dtype=object)

If you are dealing with a very large Pandas Series and if you are not sure whether an index label exists, you can check by using the ```in``` command

In [ ]:
# We check whether bananas is a food item (an index) in Groceries
x = 'bananas' in item

# We check whether bread is a food item (an index) in Groceries
y = 'chips' in item

# We print the results
print('Is bananas an index label in item:', x)
print('Is chips an index label in item:', y)

Is bananas an index label in item: False
Is chips an index label in item: True


## **Accessing and Deleting Elements in Pandas Series:**
Now let's look at how we can access or modify elements in a Pandas Series. One great advantage of Pandas Series is that it allows us to access data in many different ways. Elements can be accessed using index labels or numerical indices inside square brackets, [ ], similar to how we access elements in NumPy ndarrays. Since we can use numerical indices, we can use both positive and negative integers to access data from the beginning or from the end of the Series, respectively. 

In [ ]:
# One by their index labels
item['chocolates']

15

In [ ]:
item[['chocolates', 'milk']]   #a list of indices is passed 

chocolates    15
milk          No
dtype: object

In [ ]:
item[0]  #first element

15

In [ ]:
item[-1]  #last element

'No'

In [ ]:
item[[0,1]]    # list of numerical indices

chocolates    15
chips          5
dtype: object

In order to remove any ambiguity from whether we are reffering to the labeled index or the numerical index, panda series have two attributes <b>loc</b> and <b>iloc</b>.

* The attribute loc stands for the location and is used to explicitly state that we are using our labeled index.
* The attribute iloc stands for integer location and is used to explicitly state that we are using a numerical index

In [ ]:
item.loc[['milk', 'chips']]

milk     No
chips     5
dtype: object

In [ ]:
item.iloc[[1,2]]

chips     5
milk     No
dtype: object

Pandas Series are also mutable like NumPy ndarrays, which means we can change the elements of a Pandas Series after it has been created.

In [ ]:
# we can change the data by reassigning the value
item['milk'] = 'Yes'
item

chocolates     15
chips           5
milk          Yes
dtype: object

We can see that our series is now modified
<br><br>

## **Deleting values from Pandas Series:**

We can also delete items from a Pandas Series by using the ```.drop()``` method. The ```Series.drop(label)``` method removes the given label from the given Series.

In [ ]:
# Also we can delete items from the pandas series using drop method
item.drop('chocolates')

NameError: ignored

As we see the chocolate item is no longer is present in the series returned by the method.

><B>Note: </B>However, this drops elements from the series out of place meaning the code above just returned the modified series and actually didn't change the actual series.

In [ ]:
item   # THE CHOCOLATES ITEM MUST BE PRESENT HERE

If you want to modify the actual series, we can use either of the two methods
* overwrite the returned series on the original one

```
item = item.drop('chocolates')
```

* use inplace agrgument and set it to True

```
item.drop('chocolates', inplace=True)
```
Let's try the second one 

In [ ]:
item.drop('chocolates', inplace= True)
item

<br>

##  **Arithmatic operations on Pandas Series:**
Just like with NumPy ndarrays, we can perform element-wise arithmetic operations on Pandas Series.
We will look at the arithematic operations between pandas series and single numbers.<br>
Let's first make a new series

In [ ]:
sweets = pd.Series(data=[10, 5, 7], index=['candies', 'donuts', 'ladoos'])
sweets

candies    10
donuts      5
ladoos      7
dtype: int64

We can now modify the data in sweets by performing basic arithmetic operations. Let's see some examples

In [ ]:
sweets+2

candies    12
donuts      7
ladoos      9
dtype: int64

In [ ]:
sweets-2

candies    8
donuts     3
ladoos     5
dtype: int64

In [ ]:
sweets*2

candies    20
donuts     10
ladoos     14
dtype: int64

In [ ]:
# WE CAN ALSO APPLY MATHEMATICAL FUNCTIONS FROM NUPY SUCH AS SQAURE ROOT
import numpy as np

np.sqrt(sweets)

candies    3.162278
donuts     2.236068
ladoos     2.645751
dtype: float64

Pandas also allows us to only apply arithmetic operations on selected items in our sweets list. Let's see some examples

In [ ]:
np.power(sweets, 4)

candies    10000
donuts       625
ladoos      2401
dtype: int64

In [ ]:
np.exp(sweets)

candies    22026.465795
donuts       148.413159
ladoos      1096.633158
dtype: float64

We can also apply arithemeatic operations on specific elements

In [ ]:
sweets['candies']-4

6

In [ ]:
sweets[['donuts', 'ladoos']]*3

donuts    15
ladoos    21
dtype: int64

><b>Note:</b> We can also apply arithematic operations on a pandas series of mixed data types, provided the operations are defines on all the datatypes.

* Let's use the item series we made earlier 

In [ ]:
item = pd.Series(data=[15, 5, 'No'], index=['chocolates', 'chips', 'milk'])
item

chocolates    15
chips          5
milk          No
dtype: object

In [ ]:
item * 2  #SINCE MULTIPLICATION OPERATION IS DEFINED FOR BOTH STRINGS AND INTEGERS, THE CODE DOESN'T RETURN ANY ERROR

chocolates      30
chips           10
milk          NoNo
dtype: object

In [ ]:
item/2 #SINCE THE DIVISION OPERATION IS DEFINED ONLY FOR NUMBERS AND NOT FOR STRING, THE CODE WILL RETURN AN ERROR

TypeError: unsupported operand type(s) for /: 'str' and 'int'

<br><br><hr>
<H3><B>2. PANDAS DATAFRAME</B></H3>

Dataframe is a two dimensional object which holds rows and columns and can hold values of different data types.

* We can create a dataframe manually or by loading data from a file.

<br>
Let's first ceate a dataframe manually:<<br>

*  First let's create a dictionary of pandas series and pass it into pandas dataframe

In [ ]:
item = {'Pulkit': pd.Series([250, 15, 70, 100], index=['watch', 'toys', 'glasses', 'shirt']),
        'Ashish': pd.Series([120, 50, 90], index=['pants', 'books', 'toys' ])}

# item is a dictionary for two people containing some items and the cost of the item 

In [ ]:
# WE CAN CREATE A DATA FRAME BY PASSINF THE DICTIONARY TO THE DataFrame FUNCTION

cart = pd.DataFrame(item)
cart

Pulkit  Ashish
books       NaN    50.0
glasses    70.0     NaN
pants       NaN   120.0
shirt     100.0     NaN
toys       15.0    90.0
watch     250.0     NaN

* Make sure to capitalize the **D** and **F** while calling the dataframe function.
* The dataframe is displayed in the tabular form
* The row labels for the dataframe are built from the union of the index labels we provided in the series and the column labels for the dataframe is taken from the keys of the dictionaries.
* The dataframe has NaN values because for Pulkit we have no item like books and pants in the dicrionary we provided and similarly we have NaN values  for column Ashish.

In [ ]:
# IN ABOVE EXAMPLE WE PROVIDED THE DICTIONARIES THAT CLEARLY DEFINED THE INDEX LABLES
#HOWEVER IF WE DON'T PROVIDE THE INDEX LABELS, 
# THEN THE DATAFRAME WOULD USE THE NUMERICAL INDEX VALUES
# LET'S CREATE THE SAME DICIONARY WITHOUT THE INDEXED LABELS
new_item = {'Pulkit': pd.Series([250, 15, 70, 100]),
        'Ashish': pd.Series([120, 50, 90])}


# NOW MAKE THE DATAFRAME USING THE NEW DICTIONARIES 
df = pd.DataFrame(new_item)
df

Pulkit  Ashish
0     250   120.0
1      15    50.0
2      70    90.0
3     100     NaN

<br>
<hr>
<h4><b>2.a Attributes</b></h4>

Like we did in pandas series, we can also extract information from pandas dataframe using some attributes.

In [ ]:
cart.index

Index(['books', 'glasses', 'pants', 'shirt', 'toys', 'watch'], dtype='object')

In [ ]:
cart.columns

Index(['Pulkit', 'Ashish'], dtype='object')

In [ ]:
cart.values

array([[ nan,  50.],
       [ 70.,  nan],
       [ nan, 120.],
       [100.,  nan],
       [ 15.,  90.],
       [250.,  nan]])

In [ ]:
cart.shape

(6, 2)

In [ ]:
cart.ndim

2

In [ ]:
cart.size

12

><b>NOTE:</B> While creating the cart dataframe, we passed the whole dictionary to the dataframe function. However, there might be cases when we are only interested in some specific subset of the whole data. Pandas let's us select which data we want to put into the DataFrame, with the keywords **columns** and **index**.

In [ ]:
Pulkit_cart = pd.DataFrame(item, columns=['Pulkit'])
Pulkit_cart

Pulkit
watch       250
toys         15
glasses      70
shirt       100

In [ ]:
selected_item = pd.DataFrame(item, index=['pants', 'toys'])
selected_item

Pulkit  Ashish
pants     NaN     120
toys     15.0      90

In [ ]:
ashish_selected_item = pd.DataFrame(item, columns=['Ashish'], index=['pants', 'toys'])
ashish_selected_item

Ashish
pants     120
toys       90

We can also create a dataframe from a dictionary of lists or arrays. The procedure is same as before, we start by creating the dictionary and then pass it into the dataframe function. In this case however all the list or arrays in the dictionary must be of the same length.

In [ ]:
# Here's the dictionary of the integers and the floats.
data = {'Integers':[1,2,3],
         'Floats':[1.1, 2.2, 3.3]}

df = pd.DataFrame(data, index=['label1', 'label2', 'label3'])    ## IF WE DON'T PASS THE INDICES, DATAFRAME WILL AUTOMATICALLY USE NUMERICAL INDICES
df

Integers  Floats
label1         1     1.1
label2         2     2.2
label3         3     3.3

In [ ]:
# Creating DataFrame using a list of python dictionaries 
ListOfDict = [{'apple':20, 'banana':15, 'orange':30},{'apple':10, 'tomato':17, 'grapes': 35}]

df = pd.DataFrame(ListOfDict)
df

apple  banana  orange  tomato  grapes
0     20    15.0    30.0     NaN     NaN
1     10     NaN     NaN    17.0    35.0

DataFrame used the numerical row indices.
If we want to assign the row indices some values, we can use:


In [ ]:
df.index = ['personA', 'personB']
df  

apple  banana  orange  tomato  grapes
personA     20    15.0    30.0     NaN     NaN
personB     10     NaN     NaN    17.0    35.0

**Accessing the values in a DataFrame**

In [ ]:
df[['apple']]     ##accessing a column

apple
personA     20
personB     10

In [ ]:
df[['banana','tomato']]     ##accessing by passing a list of columns

banana  tomato
personA    15.0     NaN
personB     NaN    17.0

In [ ]:
df.loc[['personA']]    ##accessing a row

apple  banana  orange  tomato  grapes
personA     20    15.0    30.0     NaN     NaN

In [ ]:
df['grapes']['personB']    ##accessing a specific value

35.0

><b>NOTE:</B> While accessing the specific element, the column label always comes the first then the row label.

In [ ]:
## IF WE WANT TO ADD A NEW COLUMN TO OUR DATAFRAME,  WE CAN ADD LIKE THIS:

df['corn'] = [5, 7]
df

apple  banana  orange  tomato  grapes  corn
personA     20    15.0    30.0     NaN     NaN     5
personB     10     NaN     NaN    17.0    35.0     7

We can also add new columns using the arithematic operations on the other columns of our DataFrame.<br>
For eg: we can add a new column vegies by adding the values for corn and tomato


In [ ]:
df['vegies'] = df['tomato'] + df['corn']
df

apple  banana  orange  tomato  grapes  corn  vegies
personA     20    15.0    30.0     NaN     NaN     5     NaN
personB     10     NaN     NaN    17.0    35.0     7    24.0

>**NOTE:** IF YOU WANT TO ADD THE VALUES FOR ONE MORE PERSON, MEANS YOU WANT TO ADD A NEW ROW. WE FIRST HAVE TO CREATE A NEW DATAFRAME WITH THOSE ROWS AND THEN APPEND IT TO THE ORIGINAL DATAFRAME



In [ ]:
new_person = [{'apple':15, 'banana': 17, 'corn': 3, 'orange':5}]
new_df = pd.DataFrame(new_person, index=['personC'])
new_df

apple  banana  corn  orange
personC     15      17     3       5

In [ ]:
## WE CAN NOW ADD THE NEW ROW TO THE ORIGINAL DATAFRAME

df = df.append(new_df)
df

apple  banana  orange  tomato  grapes  corn  vegies
personA     20    15.0    30.0     NaN     NaN     5     NaN
personB     10     NaN     NaN    17.0    35.0     7    24.0
personC     15    17.0     5.0     NaN     NaN     3     NaN

It is also possible to insert columns in the DataFrame using **insert method** to **any location we want**.
The insert method allows us to specify:
* the location,
* label, and
* the data 

of the column that we want to add.

In [ ]:
df.insert(5, 'ginger', [7,11,13])
# THE FIRST ARGUMENT IS LOCATION OF THE COLUMN
# THE SECOND ARGUMENT IS THE LABEL FOR THE COLUMN
# AND THE LAST ONE IS THE DATA FOR THE COLUMN
df

apple  banana  orange  tomato  grapes  ginger  corn  vegies
personA     20    15.0    30.0     NaN     NaN       7     5     NaN
personB     10     NaN     NaN    17.0    35.0      11     7    24.0
personC     15    17.0     5.0     NaN     NaN      13     3     NaN

We can also delete the columns and rows using two methods:
* pop
* drop

> **NOTE:** The pop method allows us to delete columns while the drop method allows us to delete both rows and columns by using the axis keyword.

In [ ]:
df.pop('ginger')
df

apple  banana  orange  tomato  grapes  corn  vegies
personA     20    15.0    30.0     NaN     NaN     5     NaN
personB     10     NaN     NaN    17.0    35.0     7    24.0
personC     15    17.0     5.0     NaN     NaN     3     NaN

In [ ]:
df = df.drop(['grapes', 'orange'], axis=1)   # AXIS = 1 MEANS WE ARE REMOVING COLUMNS
df

apple  banana  tomato  corn  vegies
personA     20    15.0     NaN     5     NaN
personB     10     NaN    17.0     7    24.0
personC     15    17.0     NaN     3     NaN

In [ ]:
df = df.drop(['personB', 'personC'], axis=0)   # AXIS = 0 MEANS WE ARE REMOVING ROWS
df

apple  banana  tomato  corn  vegies
personA     20    15.0     NaN     5     NaN

To change the name of any row or any column,  we use the rename method



In [ ]:
df  = df.rename(columns={'banana':'pineapple'})
df

apple  pineapple  tomato  corn  vegies
personA     20       15.0     NaN     5     NaN

Now let's change the name of the row

In [ ]:
df  = df.rename(index={'personA':'A'})
df

apple  pineapple  tomato  corn  vegies
A     20       15.0     NaN     5     NaN

## **DATA CLEANING**

## **Dealing with NaN values:**
As mentioned earlier, before we can begin training our learning algorithms with large datasets, we usually need to clean the data first. This means we need to have a method for detecting and correcting errors in our data. 

While any given dataset can have many types of bad data, such as outliers or incorrect values, the type of bad data we encounter almost always is missing values. As we saw earlier, Pandas assigns ```NaN``` values to missing data. In this lesson we will learn how to detect and deal with ```NaN``` values.

We will begin by creating a DataFrame with some ```NaN``` values in it.




In [ ]:
import pandas as pd

# We create a list of Python dictionaries
items2 = [{'bikes': 20, 'pants': 30, 'watches': 35, 'shirts': 15, 'shoes':8, 'suits':45},
{'watches': 10, 'glasses': 50, 'bikes': 15, 'pants':5, 'shirts': 2, 'shoes':5, 'suits':7},
{'bikes': 20, 'pants': 30, 'watches': 35, 'glasses': 4, 'shoes':10}]

# We create a DataFrame  and provide the row index
store_items = pd.DataFrame(items2, index = ['store 1', 'store 2', 'store 3'])

# We display the DataFrame
store_items

bikes  pants  watches  shirts  shoes  suits  glasses
store 1     20     30       35    15.0      8   45.0      NaN
store 2     15      5       10     2.0      5    7.0     50.0
store 3     20     30       35     NaN     10    NaN      4.0

In cases where we load very large datasets into a DataFrame, possibly with millions of items, the number of NaN values is not easily visualized. For these cases, we can use a combination of methods to count the number of ```NaN``` values in our data. The following example combines the ```.isnull()``` and the ```sum()``` methods to count the number of ```NaN``` values in our DataFrame

In [ ]:
store_items.isnull()

bikes  pants  watches  shirts  shoes  suits  glasses
store 1  False  False    False   False  False  False     True
store 2  False  False    False   False  False  False    False
store 3  False  False    False    True  False   True    False

> In Pandas, logical True values have numerical value 1 and logical False values have numerical value 0. Therefore, we can count the number of NaN values by counting the number of logical True values.

In [ ]:
# We count the number of NaN values in the columns of store_items
x =  store_items.isnull().sum()

# We print x
print('Number of NaN values in our DataFrame:', x)

Number of NaN values in our DataFrame: bikes      0
pants      0
watches    0
shirts     1
shoes      0
suits      1
glasses    1
dtype: int64


> In order to count the total number of logical True values we use the .sum() method twice. We have to use it twice because the first sum returns a Pandas Series with the sums of logical True values along columns, as we see below:

In [ ]:
# We count the number of NaN values in store_items
x =  store_items.isnull().sum().sum()

# We print x
print('Number of NaN values in our DataFrame:', x)

Number of NaN values in our DataFrame: 3


Now that we learned how to know if our dataset has any NaN values in it, the next step is to decide what to do with them. In general we have two options, we can either delete or replace the NaN values. In the following examples we will show you how to do both.

We will start by learning how to eliminate rows or columns from our DataFrame that contain any NaN values. The .dropna(axis) method eliminates any rows with NaN values when axis = 0 is used and will eliminate any columns with NaN values when axis = 1 is used. Let's see some examples

In [ ]:
# We drop any rows with NaN values
store_items.dropna(axis = 0)

bikes  pants  watches  shirts  shoes  suits  glasses
store 2     15      5       10     2.0      5    7.0     50.0

In [ ]:
store_items

bikes  pants  watches  shirts  shoes  suits  glasses
store 1     20     30       35    15.0      8   45.0      NaN
store 2     15      5       10     2.0      5    7.0     50.0
store 3     20     30       35     NaN     10    NaN      4.0

In [ ]:
# We drop any columns with NaN values
store_items.dropna(axis = 1)

bikes  pants  watches  shoes
store 1     20     30       35      8
store 2     15      5       10      5
store 3     20     30       35     10

Notice that the .dropna() method eliminates (drops) the rows or columns with NaN values out of place. This means that the original DataFrame is not modified. You can always remove the desired rows or columns in place by setting the keyword inplace = True inside the dropna() function.

Now, instead of eliminating NaN values, we can replace them with suitable values. We could choose for example to replace all NaN values with the value 0. We can do this by using the .fillna() method as shown below.

In [ ]:
# We replace all NaN values with 0
store_items.fillna(0)

bikes  pants  watches  shirts  shoes  suits  glasses
store 1     20     30       35    15.0      8   45.0      0.0
store 2     15      5       10     2.0      5    7.0     50.0
store 3     20     30       35     0.0     10    0.0      4.0

In [ ]:
store_items.fillna(store_items.mean())

bikes  pants  watches  shirts  shoes  suits  glasses
store 1     20     30       35    15.0      8   45.0     27.0
store 2     15      5       10     2.0      5    7.0     50.0
store 3     20     30       35     8.5     10   26.0      4.0

## **READING EXTERNAL DATA**

## **Example:1**

In machine learning you will most likely use databases from many sources to train your learning algorithms. Pandas allows us to load databases of different formats into DataFrames. One of the most popular data formats used to store databases is csv. CSV stands for Comma Separated Values and offers a simple format to store data. We can load CSV files into Pandas DataFrames using the ```pd.read_csv()``` function. Let's load Google stock data into a Pandas DataFrame. The GOOG.csv file contains Google stock data from 8/19/2004 till 10/13/2017 taken from Yahoo Finance.

In [ ]:
import pandas as pd

# We load Google stock data in a DataFrame
Google_stock = pd.read_csv('GOOG.csv')

# We print some information about Google_stock
print('Google_stock is of type:', type(Google_stock))
print('Google_stock has shape:', Google_stock.shape)

Google_stock is of type: <class 'pandas.core.frame.DataFrame'>
Google_stock has shape: (3313, 7)


We see that we have loaded the stocks.csv file into a Pandas DataFrame and it consists of 3,313 rows and 7 columns. Now let's look at the stock data

In [ ]:
Google_stock

Date        Open        High         Low       Close   Adj Close  \
0     19-08-2004   49.676899   51.693783   47.669952   49.845802   49.845802   
1     20-08-2004   50.178635   54.187561   49.925285   53.805050   53.805050   
2     23-08-2004   55.017166   56.373344   54.172661   54.346527   54.346527   
3     24-08-2004   55.260582   55.439419   51.450363   52.096165   52.096165   
4     25-08-2004   52.140873   53.651051   51.604362   52.657513   52.657513   
...          ...         ...         ...         ...         ...         ...   
3308  09-10-2017  980.000000  985.424988  976.109985  977.000000  977.000000   
3309  10-10-2017  980.000000  981.570007  966.080017  972.599976  972.599976   
3310  11-10-2017  973.719971  990.710022  972.250000  989.250000  989.250000   
3311  12-10-2017  987.450012  994.119995  985.000000  987.830017  987.830017   
3312  13-10-2017  992.000000  997.210022  989.000000  989.679993  989.679993   

        Volume  
0     44994500  
1     23005800  
2     18393200  
3     15361800  
4      9257400  
...        ...  
3308    891400  
3309    968400  
3310   1693300  
3311   1262400  
3312   1157700  

[3313 rows x 7 columns]

We see that it is quite a large dataset and that Pandas has automatically assigned numerical row indices to the DataFrame. Pandas also used the labels that appear in the data in the CSV file to assign the column labels.

When dealing with large datasets like this one, it is often useful just to take a look at the first few rows of data instead of the whole dataset. We can take a look at the first 5 rows of data using the ```.head()``` method, as shown below

In [ ]:
Google_stock.head(10)

Date       Open       High        Low      Close  Adj Close    Volume
0  19-08-2004  49.676899  51.693783  47.669952  49.845802  49.845802  44994500
1  20-08-2004  50.178635  54.187561  49.925285  53.805050  53.805050  23005800
2  23-08-2004  55.017166  56.373344  54.172661  54.346527  54.346527  18393200
3  24-08-2004  55.260582  55.439419  51.450363  52.096165  52.096165  15361800
4  25-08-2004  52.140873  53.651051  51.604362  52.657513  52.657513   9257400
5  26-08-2004  52.135906  53.626213  51.991844  53.606342  53.606342   7148200
6  27-08-2004  53.700729  53.959049  52.503513  52.732029  52.732029   6258300
7  30-08-2004  52.299839  52.404160  50.675404  50.675404  50.675404   5235700
8  31-08-2004  50.819469  51.519913  50.749920  50.854240  50.854240   4954800
9  01-09-2004  51.018177  51.152302  49.512966  49.801090  49.801090   9206800

We can also take a look at the last 5 rows of data by using the ```.tail()``` method:

In [ ]:
Google_stock.tail(3)

Date        Open        High     Low       Close   Adj Close  \
3310  11-10-2017  973.719971  990.710022  972.25  989.250000  989.250000   
3311  12-10-2017  987.450012  994.119995  985.00  987.830017  987.830017   
3312  13-10-2017  992.000000  997.210022  989.00  989.679993  989.679993   

       Volume  
3310  1693300  
3311  1262400  
3312  1157700


Let's do a quick check to see whether we have any ```NaN``` values in our dataset. To do this, we will use the ```.isnull()``` method followed by the ```.any()``` method to check whether any of the columns contain ```NaN``` values.

In [ ]:
Google_stock.isnull().any()

Date         False
Open         False
High         False
Low          False
Close        False
Adj Close    False
Volume       False
dtype: bool

We see that we have no ```NaN``` values.

When dealing with large datasets, it is often useful to get statistical information from them. Pandas provides the ```.describe()``` method to get descriptive statistics on each column of the DataFrame. Let's see how this works:

In [ ]:
# We get descriptive statistics on our stock data
Google_stock.describe()

Open         High          Low        Close    Adj Close  \
count  3313.000000  3313.000000  3313.000000  3313.000000  3313.000000   
mean    380.186092   383.493740   376.519309   380.072458   380.072458   
std     223.818650   224.974534   222.473232   223.853780   223.853780   
min      49.274517    50.541279    47.669952    49.681866    49.681866   
25%     226.556473   228.394516   224.003082   226.407440   226.407440   
50%     293.312286   295.433502   289.929291   293.029114   293.029114   
75%     536.650024   540.000000   532.409973   536.690002   536.690002   
max     992.000000   997.210022   989.000000   989.679993   989.679993   

             Volume  
count  3.313000e+03  
mean   8.038476e+06  
std    8.399521e+06  
min    7.900000e+03  
25%    2.584900e+06  
50%    5.281300e+06  
75%    1.065370e+07  
max    8.276810e+07

If desired, we can apply the ```.describe()``` method on a single column as shown below:

In [ ]:
# We get descriptive statistics on a single column of our DataFrame
Google_stock['Adj Close'].describe()

count    3313.000000
mean      380.072458
std       223.853780
min        49.681866
25%       226.407440
50%       293.029114
75%       536.690002
max       989.679993
Name: Adj Close, dtype: float64

Similarly, you can also look at one statistic by using one of the many statistical functions Pandas provides. Let's look at some examples:

In [ ]:
# We print information about our DataFrame  
print()
print('Maximum values of each column:\n', Google_stock.max())
print()
print('Minimum Close value:', Google_stock['Close'].min())
print()
print('Average value of each column:\n', Google_stock.mean())


Maximum values of each column:
 Date         31-12-2015
Open              992.0
High         997.210022
Low               989.0
Close        989.679993
Adj Close    989.679993
Volume         82768100
dtype: object

Minimum Close value: 49.681866

Average value of each column:
 Open         3.801861e+02
High         3.834937e+02
Low          3.765193e+02
Close        3.800725e+02
Adj Close    3.800725e+02
Volume       8.038476e+06
dtype: float64


Another important statistical measure is data correlation. Data correlation can tell us, for example, if the data in different columns are correlated. We can use the .corr() method to get the correlation between different columns, as shown below:

In [ ]:
# We display the correlation between columns
Google_stock.corr()

Open      High       Low     Close  Adj Close    Volume
Open       1.000000  0.999904  0.999845  0.999745   0.999745 -0.564258
High       0.999904  1.000000  0.999834  0.999868   0.999868 -0.562749
Low        0.999845  0.999834  1.000000  0.999899   0.999899 -0.567007
Close      0.999745  0.999868  0.999899  1.000000   1.000000 -0.564967
Adj Close  0.999745  0.999868  0.999899  1.000000   1.000000 -0.564967
Volume    -0.564258 -0.562749 -0.567007 -0.564967  -0.564967  1.000000

A correlation value of 1 tells us there is a high correlation and a correlation of 0 tells us that the data is not correlated at all.



# **Example:2**

In [ ]:
import pandas as pd
df = pd.read_excel("excel-comp-data.xlsx")

In [ ]:
df.head()   #display first 5 rows of your data

account                         name                                street  \
0   211829   Kerluke, Koepp and Hilpert                    34456 Sean Highway   
1   320563               Walter-Trantow                     1311 Alvis Tunnel   
2   648336   Bashirian, Kunde and Price  62184 Schamberger Underpass Apt. 231   
3   109996  D'Amore, Gleichner and Bode           155 Fadel Crescent Apt. 144   
4   121213                Bauch-Goldner                   7274 Marissa Common   

              city          state  postal-code     Jan     Feb    Mar category  
0       New Jaycob          Texas        28752   10000   62000  35000        A  
1    Port Khadijah  NorthCarolina        38365   95000   45000  35000        B  
2   New Lilianland           Iowa        76517   91000  120000  35000        C  
3       Hyattburgh          Maine        46021   45000  120000  10000        A  
4  Shanahanchester     California        49681  162000  120000  35000        A

In [ ]:
df["total_sales"] = df["Jan"] + df["Feb"] + df["Mar"]
df.head()

account                         name                                street  \
0   211829   Kerluke, Koepp and Hilpert                    34456 Sean Highway   
1   320563               Walter-Trantow                     1311 Alvis Tunnel   
2   648336   Bashirian, Kunde and Price  62184 Schamberger Underpass Apt. 231   
3   109996  D'Amore, Gleichner and Bode           155 Fadel Crescent Apt. 144   
4   121213                Bauch-Goldner                   7274 Marissa Common   

              city          state  postal-code     Jan     Feb    Mar  \
0       New Jaycob          Texas        28752   10000   62000  35000   
1    Port Khadijah  NorthCarolina        38365   95000   45000  35000   
2   New Lilianland           Iowa        76517   91000  120000  35000   
3       Hyattburgh          Maine        46021   45000  120000  10000   
4  Shanahanchester     California        49681  162000  120000  35000   

  category  total_sales  
0        A       107000  
1        B       175000  
2        C       246000  
3        A       175000  
4        A       317000

In [ ]:
Jan_sales = df["Jan"].sum()
print("Total Sales of Jan is:",Jan_sales)

Total Sales of Jan is: 1462000


In [ ]:
min = df["Jan"].min()
print("Minimum Sales of Jan is:",min)

Minimum Sales of Jan is: 10000


In [ ]:
max = df["Jan"].max()
print("Minimum Sales of Jan is:",max)

Minimum Sales of Jan is: 162000


In [ ]:
avg_Jan = df["Jan"].mean()
print("Minimum Sales of Jan is: $",round(avg_Jan,2))

Minimum Sales of Jan is: $ 97466.67


In [ ]:
sorted_ascending = df.sort_values("total_sales")
sorted_ascending

account                              name  \
0    211829        Kerluke, Koepp and Hilpert   
1    320563                    Walter-Trantow   
3    109996       D'Amore, Gleichner and Bode   
8    209744                    Champlin-Morar   
13   268755                       Walsh-Haley   
10   214098       Goodwin, Homenick and Jerde   
9    212303                    Gerhold-Maggio   
2    648336        Bashirian, Kunde and Price   
6    145068                        Casper LLC   
7    205217                  Kovacek-Johnston   
5    132971  Williamson, Schumm and Hettinger   
4    121213                     Bauch-Goldner   
12   242368      Frami, Anderson and Donnelly   
11   231907                        Hahn-Moore   
14   273274                     McDermott PLC   

                                  street               city          state  \
0                     34456 Sean Highway         New Jaycob          Texas   
1                      1311 Alvis Tunnel      Port Khadijah  NorthCarolina   
3            155 Fadel Crescent Apt. 144         Hyattburgh          Maine   
8                       26739 Grant Lock    Lake Juliannton   Pennsylvania   
13                  2624 Beatty Parkways       Goodwinmouth    RhodeIsland   
10             649 Cierra Forks Apt. 078           Rosaberg       Tenessee   
9              366 Maggio Grove Apt. 998          North Ras          Idaho   
2   62184 Schamberger Underpass Apt. 231     New Lilianland           Iowa   
6           340 Consuela Bridge Apt. 400  Lake Gabriellaton     Mississipi   
7           91971 Cronin Vista Suite 601         Deronville    RhodeIsland   
5                   89403 Casimer Spring       Jeremieburgh       Arkansas   
4                    7274 Marissa Common    Shanahanchester     California   
12                       182 Bertie Road        East Davian           Iowa   
11              18115 Olivine Throughway      Norbertomouth    NorthDakota   
14                 8917 Bergstrom Meadow    Kathryneborough       Delaware   

    postal-code     Jan     Feb     Mar category  total_sales  
0         28752   10000   62000   35000        A       107000  
1         38365   95000   45000   35000        B       175000  
3         46021   45000  120000   10000        A       175000  
8         64415   70000   95000   35000        C       200000  
13        31919   55000  120000   35000        C       210000  
10        47743   45000  120000   55000        A       220000  
9         46308   70000  120000   35000        C       225000  
2         76517   91000  120000   35000        C       246000  
6         18008   62000  120000   70000        B       252000  
7         53461  145000   95000   35000        A       275000  
5         62785  150000  120000   35000        A       305000  
4         49681  162000  120000   35000        A       317000  
12        72686  162000  120000   35000        B       317000  
11        31415  150000   10000  162000        B       322000  
14        27933  150000  120000   70000        A       340000

In [ ]:
sorted_d = df.sort_values("total_sales", ascending = False)
sorted_d.head()

account                              name                    street  \
14   273274                     McDermott PLC     8917 Bergstrom Meadow   
11   231907                        Hahn-Moore  18115 Olivine Throughway   
4    121213                     Bauch-Goldner       7274 Marissa Common   
12   242368      Frami, Anderson and Donnelly           182 Bertie Road   
5    132971  Williamson, Schumm and Hettinger      89403 Casimer Spring   

               city        state  postal-code     Jan     Feb     Mar  \
14  Kathryneborough     Delaware        27933  150000  120000   70000   
11    Norbertomouth  NorthDakota        31415  150000   10000  162000   
4   Shanahanchester   California        49681  162000  120000   35000   
12      East Davian         Iowa        72686  162000  120000   35000   
5      Jeremieburgh     Arkansas        62785  150000  120000   35000   

   category  total_sales  
14        A       340000  
11        B       322000  
4         A       317000  
12        B       317000  
5         A       305000

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   account      15 non-null     int64 
 1   name         15 non-null     object
 2   street       15 non-null     object
 3   city         15 non-null     object
 4   state        15 non-null     object
 5   postal-code  15 non-null     int64 
 6   Jan          15 non-null     int64 
 7   Feb          15 non-null     int64 
 8   Mar          15 non-null     int64 
 9   category     15 non-null     object
 10  total_sales  15 non-null     int64 
dtypes: int64(6), object(5)
memory usage: 1.4+ KB


Let's compare the sales for each month and find which month has highest sale

In [ ]:
def max_sales(row):
    if row['Jan']>row['Feb'] and  row['Jan']>row['Mar']:
        return("Jan sale is high")
    elif row['Feb']>row['Jan'] and  row['Feb']>row['Mar']:
        return("Feb sale is high")
    else:
        return("Mar sale is high")

df['max'] = df.apply(max_sales, axis = 1) #axis = 1 means that the function is applies to each row

In [ ]:
df.head()

account                         name                                street  \
0   211829   Kerluke, Koepp and Hilpert                    34456 Sean Highway   
1   320563               Walter-Trantow                     1311 Alvis Tunnel   
2   648336   Bashirian, Kunde and Price  62184 Schamberger Underpass Apt. 231   
3   109996  D'Amore, Gleichner and Bode           155 Fadel Crescent Apt. 144   
4   121213                Bauch-Goldner                   7274 Marissa Common   

              city          state  postal-code     Jan     Feb    Mar  \
0       New Jaycob          Texas        28752   10000   62000  35000   
1    Port Khadijah  NorthCarolina        38365   95000   45000  35000   
2   New Lilianland           Iowa        76517   91000  120000  35000   
3       Hyattburgh          Maine        46021   45000  120000  10000   
4  Shanahanchester     California        49681  162000  120000  35000   

  category  total_sales               max  
0        A       107000  Feb sale is high  
1        B       175000  Jan sale is high  
2        C       246000  Feb sale is high  
3        A       175000  Feb sale is high  
4        A       317000  Jan sale is high

In [ ]:
for x in df:
    print(x)

account
name
street
city
state
postal-code
Jan
Feb
Mar
category
total_sales
max


In [ ]:
#for loop to get the address which consist of city, state and postal code 
for x in df.index: 
     print(df['city'][x] +", "+ df['state'][x]+", "+str(df['postal-code'][x]))

New Jaycob, Texas, 28752
Port Khadijah, NorthCarolina, 38365
New Lilianland, Iowa, 76517
Hyattburgh, Maine, 46021
Shanahanchester, California, 49681
Jeremieburgh, Arkansas, 62785
Lake Gabriellaton, Mississipi, 18008
Deronville, RhodeIsland, 53461
Lake Juliannton, Pennsylvania, 64415
North Ras, Idaho, 46308
Rosaberg, Tenessee, 47743
Norbertomouth, NorthDakota, 31415
East Davian, Iowa, 72686
Goodwinmouth, RhodeIsland, 31919
Kathryneborough, Delaware, 27933


In [ ]:
# We display the total sales for each category using group-by
df.groupby(['category'])['total_sales'].sum()

category
A    1739000
B    1066000
C     881000
Name: total_sales, dtype: int64

# CASE STUDY: ANALYSING GOOGLE PLAY STORE DATASET

# To predict the rating for a mobile app given features like size, number of downloads, etc.

# Data Description

## Attributes

● **App**: Application name

● **Category**: Category the app belongs to

● **Rating**: Overall user rating of the app

● **Reviews**: Number of user reviews for the app

● **Size**: Size of the app

● **Installs**: Number of user downloads/installs for the app

● **Type**: Paid or Free

● **Price**: Price of the app

● **Content Rating**: Age group the app is targeted at - Children / Mature 21+ / Adult

● **Genres**: An app can belong to multiple genres (apart from its main category). For
eg, a musical family game will belong to Music, Game, Family genres.

● **Last Updated**: Date when the app was last updated on Play Store

● **Current** Ver: Current version of the app available on Play Store

● **Android Ver**: Min required Android version


##  Install all the necessary libraries and read the provided dataset.

In [ ]:
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv("Apps_data.csv")
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Rating_cat Reviews  Size     Installs  Type Price Content Rating  \
0       High     159   19M      10,000+  Free     0       Everyone   
1       High     967   14M     500,000+  Free     0       Everyone   
2       High   87510  8.7M   5,000,000+  Free     0       Everyone   
3       High  215644   25M  50,000,000+  Free     0           Teen   
4       High     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

##  Check the info and summary statistics of the dataset.

In [ ]:
df.info()

NameError: ignored

In [ ]:
df.isna().sum()

There are some missing values in Rating, Type, Content Rating, Current ver, Android Ver.

Next, Check if there are any duplicate entries for the apps. 

## Check for duplicates

In [ ]:
print(df.App.value_counts())

ROBLOX                                                9
CBS Sports App - Scores, News, Stats & Watch Live     8
ESPN                                                  7
Duolingo: Learn Languages Free                        7
Candy Crush Saga                                      7
                                                     ..
Meet U - Get Friends for Snapchat, Kik & Instagram    1
U-Report                                              1
U of I Community Credit Union                         1
Waiting For U Launcher Theme                          1
iHoroscope - 2018 Daily Horoscope & Astrology         1
Name: App, Length: 9660, dtype: int64


In [ ]:
df = df.drop_duplicates(subset=['App'])
df.duplicated(subset=['App']).sum()

0

In [ ]:
#re-check duplicate apps
print(df.App.value_counts())

Photo Editor & Candy Camera & Grid & ScrapBook    1
CK Pharmacies                                     1
CK Call NEW                                       1
Ck Coif                                           1
CONTRACT KILLER: ZOMBIES                          1
                                                 ..
Bike Race Free - Top Motorcycle Racing Games      1
Dance School Stories - Dance Dreams Come True     1
3D Bowling                                        1
Mind Games Pro                                    1
iHoroscope - 2018 Daily Horoscope & Astrology     1
Name: App, Length: 9660, dtype: int64


##  Check if there are any wrong values in the ‘Category’ column

In [ ]:
print(df.Category.value_counts())

FAMILY                 1832
GAME                    959
TOOLS                   827
BUSINESS                420
MEDICAL                 395
PERSONALIZATION         376
PRODUCTIVITY            374
LIFESTYLE               369
FINANCE                 345
SPORTS                  325
COMMUNICATION           315
HEALTH_AND_FITNESS      288
PHOTOGRAPHY             281
NEWS_AND_MAGAZINES      254
SOCIAL                  239
BOOKS_AND_REFERENCE     222
TRAVEL_AND_LOCAL        219
SHOPPING                202
DATING                  171
VIDEO_PLAYERS           163
MAPS_AND_NAVIGATION     131
EDUCATION               119
FOOD_AND_DRINK          112
ENTERTAINMENT           102
AUTO_AND_VEHICLES        85
LIBRARIES_AND_DEMO       84
WEATHER                  79
HOUSE_AND_HOME           74
ART_AND_DESIGN           64
EVENTS                   64
PARENTING                60
COMICS                   56
BEAUTY                   53
1.9                       1
Name: Category, dtype: int64


**‘Category’** has an invalid entry which is **1.9**. Category **FAMILY** has the highest number of apps.
We shall replace entry **1.9** and impute that with most occuring entry which is **Family**

In [ ]:
df['Category'].replace("1.9", "FAMILY", inplace=True)
df['Category'].value_counts()

FAMILY                 1833
GAME                    959
TOOLS                   827
BUSINESS                420
MEDICAL                 395
PERSONALIZATION         376
PRODUCTIVITY            374
LIFESTYLE               369
FINANCE                 345
SPORTS                  325
COMMUNICATION           315
HEALTH_AND_FITNESS      288
PHOTOGRAPHY             281
NEWS_AND_MAGAZINES      254
SOCIAL                  239
BOOKS_AND_REFERENCE     222
TRAVEL_AND_LOCAL        219
SHOPPING                202
DATING                  171
VIDEO_PLAYERS           163
MAPS_AND_NAVIGATION     131
EDUCATION               119
FOOD_AND_DRINK          112
ENTERTAINMENT           102
AUTO_AND_VEHICLES        85
LIBRARIES_AND_DEMO       84
WEATHER                  79
HOUSE_AND_HOME           74
EVENTS                   64
ART_AND_DESIGN           64
PARENTING                60
COMICS                   56
BEAUTY                   53
Name: Category, dtype: int64

##  Work on **Ratings** column

In [ ]:
df['Rating'].isna().sum()  #check for missing values

1474

Rating column has 1463 missing values. We shall fill with mean

In [ ]:
df['Rating'].fillna(df['Rating'].mean(), inplace=True)
df['Rating'].isna().sum()

0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          10841 non-null  float64
 3   Rating_cat      10841 non-null  object 
 4   Reviews         10841 non-null  object 
 5   Size            10841 non-null  object 
 6   Installs        10841 non-null  object 
 7   Type            10840 non-null  object 
 8   Price           10841 non-null  object 
 9   Content Rating  10840 non-null  object 
 10  Genres          10841 non-null  object 
 11  Last Updated    10841 non-null  object 
 12  Current Ver     10833 non-null  object 
 13  Android Ver     10838 non-null  object 
dtypes: float64(1), object(13)
memory usage: 1.2+ MB


##  Make the values of **‘Size’** as integers by replacing **M** and **K** with correct values. Replace **“Varies with device”** with **200000** and convert all the values to numeric.

In [ ]:
df["Size"] = df["Size"].str.replace('Varies with device', '200000')
df["Size"] = df["Size"].str.replace('M', '000000')
df["Size"] = df["Size"].str.replace('K', '000')
df["Size"] = pd.to_numeric(df["Size"], errors = 'coerce')    #It will replace all non-numeric values with NaN.
df["Size"] = df["Size"].fillna(df["Size"].mean())

In [ ]:
df.info()

##  Remove “,” and “+” from the values of the “Installs” column and change the datatype.

In [ ]:
df["Installs"] = df["Installs"].str.replace(',', '')
df["Installs"] = df["Installs"].str.replace('+', '')
df["Installs"] = pd.to_numeric(df["Installs"], errors = 'coerce')

In [ ]:
df.info()

##  What is the percentage of paid apps in the data?

In [ ]:
df["Type"].value_counts()

In [ ]:
#Since there is a type of 0 , we shall replace it with Free
df["Type"].replace('0', 'Free', inplace=True)

In [ ]:
df["Type"].value_counts()

In [ ]:
Paid_apps = df[df.Type == 'Paid'].shape[0]
Free_apps = df[df.Type == 'Free'].shape[0]
print("Paid App Percentage=",round((Paid_apps * 100)/(Paid_apps + Free_apps),2),"%")

##  Remove the $ sign the “Price” column values and make it a numerical column.

In [ ]:
df["Price"] = df["Price"].str.replace('$', '')
df["Price"] = pd.to_numeric(df["Price"], errors = 'coerce')

In [ ]:
df.info()

##  The most expensive app and how much does it cost

In [ ]:
expensive_ap = df.sort_values(by ='Price', ascending=False )
expensive_ap.head(1)

Most expensive app is **I'm Rich - Trump Edition** and it costs **400$**

##  Drop columns that you feel cannot be used for model building. Example-App, Content
Rating, Genre, Last updated etc. from the final data frame

In [ ]:
df.drop(['App', 'Content Rating', 'Genres', "Last Updated", "Current Ver", "Android Ver"], axis = 1, inplace=True)
df.head()

NameError: ignored